### ISSUE TITLE
About the Dice loss

### ISSUE THREAD

**Issue by fepegar** [NifTK/NiftyNet#128]
The default loss function according to the [config docs](http://niftynet.readthedocs.io/en/dev/config_spec.html#training) seems to be "Dice", which seems to be [`niftynet.layer.loss_segmentation.dice`](https://github.com/NifTK/NiftyNet/blob/dev/niftynet/engine/application_factory.py#L81-L82) which, according to the [docstring](https://github.com/NifTK/NiftyNet/blob/dev/niftynet/layer/loss_segmentation.py#L378-L386), is the loss function described in [Milletari et al.](https://arxiv.org/pdf/1606.04797.pdf).

The implementation in NiftyNet [squares](https://github.com/NifTK/NiftyNet/blob/dev/niftynet/layer/loss_segmentation.py#L413) the prediction in the denominator, but it seems that in Milletari's [implementation](https://github.com/faustomilletari/VNet/blob/master/pyLayer.py#L36-L40), the prediction is a binary array, so the squaring is just the total number of positives. The paper says "*where the sums run over the N voxels, of the predicted binary segmentation
volume pi ∈ P and the ground truth binary volume gi ∈ G*".

So it seems to me that the default loss is different to the one described in the paper. Is that correct?

**Comment by wyli** [comment]
Hi @fepegar yes you're right, by default the predicted probabilities are not quantised to discrete values before computing the loss. This is actually consistent with Milletari's implementation (https://github.com/faustomilletari/VNet/blob/master/pyLayer.py#L62).

**Comment by fepegar** [comment]
Hi @wyli. I'm not sure what you mean. This is his code for the loss:
```python
dice = np.zeros(bottom[0].data.shape[0],dtype=np.float32)
self.union = np.zeros(bottom[0].data.shape[0],dtype=np.float32)
self.intersection = np.zeros(bottom[0].data.shape[0],dtype=np.float32)

self.result = np.reshape(np.squeeze(np.argmax(bottom[0].data[...],axis=1)),[bottom[0].data.shape[0],bottom[0].data.shape[2]])
self.gt = np.reshape(np.squeeze(bottom[1].data[...]),[bottom[1].data.shape[0],bottom[1].data.shape[2]])

self.gt = (self.gt > 0.5).astype(dtype=np.float32)
self.result = self.result.astype(dtype=np.float32)

for i in range(0,bottom[0].data.shape[0]):
    # compute dice
    CurrResult = (self.result[i,:]).astype(dtype=np.float32)
    CurrGT = (self.gt[i,:]).astype(dtype=np.float32)

    self.union[i]=(np.sum(CurrResult) + np.sum(CurrGT))
    self.intersection[i]=(np.sum(CurrResult * CurrGT))

    dice[i] = 2 * self.intersection[i] / (self.union[i]+0.00001)
```

Note how the logits (`self.result`) are integer values, I guess 0s and 1s, coming from `np.argmax`:
```python
# (More readable version)
argmax = np.argmax(bottom[0].data, axis=1)
self.result = np.squeeze(argmax).squeeze()
self.result = np.reshape(self.result, [bottom[0].data.shape[0],
                                       bottom[0].data.shape[2]])
```
and the ground truth is binarised by thresholding:
```python
self.gt = (self.gt > 0.5).astype(dtype=np.float32)
```


In NiftyNet:
```python
dice_denominator = \
    tf.reduce_sum(tf.square(prediction), reduction_indices=[0]) + \
    tf.sparse_reduce_sum(one_hot, reduction_axes=[0])
```

`dice_denominator` of NiftyNet is equivalent to `self.union[i]` in VNet, so maybe `prediction` should be thresholded and the `tf.square` should be removed since it would have no effect :
```python
dice_denominator = \
    tf.reduce_sum(prediction, reduction_indices=[0]) + \
    tf.sparse_reduce_sum(one_hot, reduction_axes=[0])
```

**Comment by wyli** [comment]
Hi @fepegar `self.result` is not used in backpropagation. Anyway this is now out of scope of this issue list...

**Comment by fepegar** [comment]
Why is this out of scope? I think it's a bug in the computation of the loss, that gives slightly different results if fixed.

In VNet backprop, `self.union` from the forward pass is used, which comes from `CurrResult`, which comes from `self.result`.

**Comment by wyli** [comment]
@fepegar thanks for raising the issue and the discussions, to conclude -- NiftyNet's implementation is not exactly following Milletari's implementation; and both implementations are not exactly following the paper... But given that the discrepancies are small and NiftyNet's version works well in many applications let's keep it as default and always open to better solutions.

**Comment by fepegar** [comment]
As a reference, here's some theoretical justification from @faustomilletari, thesis (page number 72):
https://mediatum.ub.tum.de/doc/1395260/1395260.pdf

**Comment by argman** [comment]
@wyli , why do you use square in prediction ?
